shakenfist/virt.py

```diff
@@ -53,39 +53,82 @@ def _safe_int_cast(i):
+            {
+                'type': 'raw',
+                'device': config_device,
+                'bus': bus,
+                'path': os.path.join(instance_path, config_device),
+                'present_as': 'disk',
+                'snapshot_ignores': True
+            }
+        ]
+    }
+
+    i = 0
+    for d in disk_spec[1:]:
+        bus = _get_defaulted_disk_bus(d)
+        device = _get_disk_device_base(bus) + chr(ord('c') + i)
+        block_devices['devices'].append({
+            'type': disk_type,
+            'size': _safe_int_cast(d.get('size')),
+            'device': device,
+            'bus': bus,
+            'path': os.path.join(instance_path, device),
+            'base': d.get('base'),
+            'present_as': _get_defaulted_disk_type(d),
+            'snapshot_ignores': False
+        })
+        i += 1
+
+    block_devices['finalized'] = False
+    return block_devices
+
+
+def instance_path(instance_uuid):
+    return os.path.join(config.get('STORAGE_PATH'), 'instances', instance_uuid)
+
+
+def _snapshot_path(instance_uuid):
+    return os.path.join(config.get('STORAGE_PATH'), 'snapshots')
+
+
+def _xml_file(instance_uuid):
+    return os.path.join(instance_path(instance_uuid), 'libvirt.xml')
+
+
 class Instance(object):
-    def __init__(self, cpus, memory, name, namespace, uuid,
-                 block_devices=None, devices=None, console_port=0,
-                 disk_spec=None, error_message=None, node=None,
-                 node_history=None, placement_attempts=0, power_state='',
-                 power_state_previous=None, power_state_updated=0,
-                 requested_placement=None, ssh_key=None, state=None,
-                 state_updated=None, user_data=None, vdi_port=0,
-                 video=None):
-
-        self.uuid = uuid
-        self.memory = memory
-        self.name = name
-        self.namespace = namespace
-        self.block_devices = block_devices
-        self.devices = devices
-        self.console_port = console_port
-        self.cpus = cpus
-        self.disk_spec = disk_spec
-        self.error_message = error_message
-        self.node = node
-        self.node_history = node_history
-        self.placement_attempts = placement_attempts
-        self.power_state = power_state
-        self.power_state_previous = power_state_previous
-        self.power_state_updated = power_state_updated
-        self.requested_placement = requested_placement
-        self.ssh_key = ssh_key
-        self.state = state
-        self.state_updated = state_updated
-        self.user_data = user_data
-        self.vdi_port = vdi_port
-        self.video = video
+    def __init__(self, static_values):
+        # This dictionary contains values which will never change for this
+        # instance. It is therefore safe to cache forever. I know this moves
+        # us away from member variables again, but I wanted to make it very
+        # clear that these values are not meant to be assigned to.
```

Comment:
These static values complicate (look non-pretty) the use of the class. We can enforce the read-only nature using properties.